### PR TITLE
[FEATURE: UIPT-5]  - Application router

### PR DESCRIPTION
**Pull request by TorskyM**

### CODE REVIEW

File: src/app/app-routing.module.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { NgModule }             from '@angular/core';
+import { RouterModule, Routes } from '@angular/router';
+
+import { NotFoundComponent }        from './pages/not-found/not-found.component';
+import { HomeComponent }            from './pages/home/home.component';
+import { ProductListPageComponent } from './pages/product-list-page/product-list-page.component';
+import { ProductDetailsPageComponent } from './pages/product-details-page/product-details-page.component';
+import { ProductResolver }          from './product.resolver';
```
Comment (AlexLarusso) [inline src/app/app-routing.module.ts]: please align

File: src/app/app-routing.module.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  { path: 'products/:id', component: ProductDetailsPageComponent, resolve: { products: ProductResolver } },
+  { path: '404',          component: NotFoundComponent },
+  { path: '**',           redirectTo: '/404' },
+]
+
+@NgModule({
+  imports: [ RouterModule.forRoot(routes) ],
+  exports: [ RouterModule ]
+})
+export class AppRoutingModule { }
```
Comment (AlexLarusso) [inline src/app/app-routing.module.ts]: Don't use separate module for it.

File: src/app/pages/not-found/not-found.component.html [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<app-header></app-header>
+<div class="not-found">
+<h1 class="not-found__title">Looks like it`s nothing here...</h1>
+</div>
```
Comment (AlexLarusso) [inline src/app/pages/not-found/not-found.component.html]: Align please.

Don't name views with `component`. Just `not-found.html`

Check everywhere

File: src/app/product-details/product-details.component.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+import { ActivatedRoute } from '@angular/router';
+
+@Component({
+  selector: 'app-product-details',
+  templateUrl: './product-details.component.html',
+  styleUrls: ['./product-details.component.scss']
+})
+export class ProductDetailsComponent {
+ id = Number(this.route.snapshot.paramMap.get('id'));
```
Comment (AlexLarusso) [inline src/app/product-details/product-details.component.ts]: Please define that id is `public`

File: src/app/product.resolver.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import { ActivatedRoute } from '@angular/router';
+
+@Injectable()
+export class ProductResolver implements Resolve<IProduct> {
+  
+  constructor(private route: ActivatedRoute, private ProductService: ProductService) {}
+  
+  resolve(route: ActivatedRouteSnapshot): Observable<IProduct> {
+    return this.ProductService.getProduct(Number(route.paramMap.get('id')))
+    .pipe(take(1), map((product: IProduct) => product));
```
Comment (AlexLarusso) [inline src/app/product.resolver.ts]: align.
`
map((product: IProduct) => product)
`

You do nothing here. Remove this logic

File: src/app/product.service.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import { map } from 'rxjs/operators';
+
+import { IProduct } from './product';
+import { HttpService } from './http.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ProductService {
+  constructor(private HttpService: HttpService){}
```
Comment (AlexLarusso) [inline src/app/product.service.ts]: For instances use lowercase `httpService`

File: src/app/shared/main-logo/main-logo.component.html [hunk 7]
```diff
@@ -1,6 +1,8 @@
-<div class="logo" [ngClass]="{
+<a routerLink="/home">
+  <div  class="logo" [ngClass]="{
```
Comment (AlexLarusso) [inline src/app/shared/main-logo/main-logo.component.html]: remove useless space

File: src/app/shared/product-item/product-item.component.html [hunk 8]
```diff
@@ -1,10 +1,12 @@
 <div class="col-3 product-item__container">
   <div class="product-item__picture__container">
+    <a routerLink="/products/{{ id }}">
     <img class="product-item__picture"
       class="product-item__picture"
       [alt]="title"
       [src]="src"
     />
```
Comment (AlexLarusso) [inline src/app/shared/product-item/product-item.component.html]: align

File: src/app/app.module.ts [hunk 9]
```diff
@@ -27,6 +27,22 @@ import { ProductShortInfoService } from './product-short-info.service';
 import { ScrollAnchorDirective } from './scroll-anchor.directive';
 import { ScrollComponent } from './scroll/scroll.component';
 import { ProductCarouselComponent } from './shared/product-carousel/product-carousel.component';
+import { NotFound }                from './pages/not-found/not-found';
+import { Home }                    from './pages/home/home';
+import { ProductListPage }         from './pages/product-list-page/product-list-page';
+import { ProductDetailsComponent }  from './product-details/product-details.component';
+import { ProductDetailsPage }      from './pages/product-details-page/product-details-page';
+import { ProductResolver }         from './product.resolver';
+import { Routes, RouterModule } from '@angular/router';
```
Comment (AlexLarusso) [inline src/app/app.module.ts]: Please align everything.

File: src/app/http.service.ts [hunk 10]
```diff
@@ -6,9 +6,9 @@ import { Observable } from 'rxjs';
 @Injectable()
 export class HttpService {
   private URL = '../assets/server-data/data.json';
-
+
```
Comment (AlexLarusso) [inline src/app/http.service.ts]: Remove trailing spaces

File: src/app/pages/home/home.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+export class Home {
+  @ViewChildren(ScrollAnchorDirective)
+    private pageAnchors: QueryList<ScrollAnchorDirective>;
+    
+  constructor(private scrollService: ScrollService) {}
+
+  ngAfterViewInit () {
+    this.pageAnchors.forEach(el => 
+      this.scrollService.addAnchor(el.elementReference)
+    );
```
Comment (AlexLarusso) [inline src/app/pages/home/home.ts]: One line please

File: src/app/pages/home/home.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  selector: 'app-home',
+  templateUrl: './home.html',
+})
+export class Home {
+  @ViewChildren(ScrollAnchorDirective)
+    private pageAnchors: QueryList<ScrollAnchorDirective>;
+    
+  constructor(private scrollService: ScrollService) {}
+
+  ngAfterViewInit () {
```
Comment (AlexLarusso) [inline src/app/pages/home/home.ts]: Typescript!!
`
public ngAfterViewInit(): void {...
`

File: src/app/pages/not-found/not-found.html [hunk 13]
```diff
@@ -0,0 +1,5 @@
+<app-header></app-header>
+<div class="not-found">
+<h1 class="not-found__title">Looks like it`s nothing here...</h1>
```
Comment (AlexLarusso) [inline src/app/pages/not-found/not-found.html]: Add 2 spaces

File: src/app/product-details/product-details.component.ts [hunk 14]
```diff
@@ -0,0 +1,15 @@
+import { Component } from '@angular/core';
+
+import { ActivatedRoute } from '@angular/router';
+
+@Component({
+  selector: 'app-product-details',
+  templateUrl: './product-details.component.html',
```
Comment (AlexLarusso) [inline src/app/product-details/product-details.component.ts]: Rename all html files:
`product-details.component.html` -> `product-details.html`

File: src/app/product-details/product-details.component.ts [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+  styleUrls: ['./product-details.component.scss']
```
Comment (AlexLarusso) [inline src/app/product-details/product-details.component.ts]: Rename all scss files:
`
product-details.component.scss -> product-details.scss
`

File: src/app/product-details/product-details.component.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+})
+export class ProductDetailsComponent {
+  public id = Number(this.route.snapshot.paramMap.get('id'));
+
+ constructor(private route: ActivatedRoute ){}
```
Comment (AlexLarusso) [inline src/app/product-details/product-details.component.ts]: remove useless dependencie

File: src/app/product-list/product-list.component.html [hunk 17]
```diff
@@ -20,7 +20,8 @@ <h2 class="col-12 content__title">From Our Collection</h2>
         class="product__item"
         [src]="product.imgUrl" 
         [title]="product.title"
-        [price]="product.price">
+        [price]="product.price"
+        [id]="product.id">
```
Comment (AlexLarusso) [inline src/app/product-list/product-list.component.html]: Don't use reserved attributes to bind;

`
[id] -> [productId]
`

File: src/app/product.resolver.ts [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+import { ActivatedRouteSnapshot } from '@angular/router';
+
+import { IProduct } from './product';
+import { ProductService } from './product.service';
+import { ActivatedRoute } from '@angular/router';
+
+@Injectable()
+export class ProductResolver implements Resolve<IProduct> {
+
+  constructor(private route: ActivatedRoute, private ProductService: ProductService) { }
```
Comment (AlexLarusso) [inline src/app/product.resolver.ts]: All instances should be in lowercase:

`private ProductService` -> `private productService`

File: src/app/product-details/product-details.component.ts (hunk 14 above)
Comment (TorskyM) [reply src/app/product-details/product-details.component.ts]: Can we do this in the next sprint?

**Pull request merged**